tapdb: remove duplicate assets before adding unique index

guggero opened the pull request:
This commit first removes duplicate assets that were mistakenly created due to self transfers in previous versions.
After removing the duplicates, the new unique index should be applied without issues.

gijswijs commented on tapdb/sqlc/migrations/000020_asset_unique_key.up.sql:
```diff
@@ -1,3 +1,43 @@
+-- First step is to make sure all duplicate assets are seen as spent, if their
+-- anchor transaction is spent. That way we can later also group on the spent
+-- column to remove duplicates.
+UPDATE assets
```
I'm not sure if I grok this first query. How are we setting all duplicate assets here as being spent? What is it in the CTE that selects for duplicates exclusively?

gijswijs replied:
Because the duplication happened during address self-transfer all assets should be considered spent in the first place?

guggero replied:
No, the logic that updates the `spent` flag in the normal application flow will only do that for the "original" assets, not the duplicates. So in order for us to be able to properly de-duplicate the assets, we'll just make sure the `spent` flag is set correctly for all of them. Update the comment on the query as well.

gijswijs commented on tapdb/sqlc/migrations/000020_asset_unique_key.up.sql:
```diff
@@ -1,3 +1,43 @@
+
+-- Now remove all asset witnesses for the assets we're going to remove in the
+-- next step.
+DELETE
```
Can we safely delete the witness of a duplicate asset? Are we sure the witness is also available for the asset that will remain? Otherwise we should UPDATE the foreign key of the asset witness.

guggero replied:
Yes, I think so. We always insert the witness with the asset ID that was inserted in the previous step. So every asset should have a witness.

guggero replied:
Update the logic to make sure we fill any gaps for duplicates that for some reason didn't have a witness.

gijswijs reviewed (commented):
There are four tables with a foreign key reference to `assets`:
- `asset_witnesses`
- `asset_proofs`
- `passive_assets`
- `addr_events`

This PR addresses `asset_witnesses` but ignores the other three tables. Are we sure we can safely do so?

guggero commented:
> There are four tables with a foreign key reference to `assets`:
> 
>     * `asset_witnesses`
> 
>     * `asset_proofs`
> 
>     * `passive_assets`
> 
>     * `addr_events`
> 
> 
> This PR addresses `asset_witnesses` but ignores the other three tables. Are we sure we can safely do so?

Yeah, good point. In the example database I was given there were no entries in the `asset_proofs`, `passive_assets` and `addr_events` table. So I think due to how the other insert/upsert queries are structured, only the asset with the lowest ID would get an entry in those tables. So by keeping the asset with the lowest ID when removing duplicates should work.
But I'll look into some more safe queries with the help of LLMs.

guggero commented:
@gijswijs I added some more involved logic to properly account for what to do with the `asset_witnesses` and `asset_proofs` table. Those should now be dealt with correctly, I also added test code for that.

That leaves the `passive_assets` and `addr_events` tables. I thought hard about the code that produced the duplicates and am 95% sure that we won't ever have entries in those two tables for the duplicate entries. And if for some reason there would be, then for those users the migration would fail and we could manually intervene.

ffranr commented on tapdb/migrations_test.go:
```diff
@@ -130,3 +132,64 @@ func TestMigrationDowngrade(t *testing.T) {
 	require.ErrorIs(t, err, ErrMigrationDowngrade)
 }
+
+func TestMigration20(t *testing.T) {
```
we should add a doc. It might describe on a high level which aspects of the version 20 migration we're testing.

ffranr commented on tapdb/migrations_test.go:
```diff
@@ -130,3 +132,64 @@ func TestMigrationDowngrade(t *testing.T) {
+	err := db.ExecuteMigrations(TargetLatest)
+	require.NoError(t, err)
+
+	_, assetStore := newAssetStoreFromDB(db.BaseDB)
```
At this point in the test, we are primarily verifying the state of the database after it has been loaded. We should add a comment to clarify this.

ffranr commented on tapdb/sqlc/migrations/000020_asset_unique_key.up.sql:
```diff
@@ -1,4 +1,107 @@
+                                 ON a.anchor_utxo_id = mu.utxo_id
+                            JOIN chain_txns ct
+                                 ON mu.txn_id = ct.txn_id
+                            LEFT JOIN asset_transfer_inputs ati
+                                      ON ati.anchor_point = mu.outpoint
+                   WHERE a.spent = false
+                     AND ati.input_id IS NOT NULL);
```
Please correct me if my understanding is wrong: if `ati.input_id IS NOT NULL` then the asset is spent? And those are the rows that we ensure are set to spent, is that correct?

guggero replied:
Correct. We `LEFT JOIN` here, meaning that any field on `ati` is `NULL` if there is no matching spending transaction referencing the asset's on-chain outpoint.

ffranr commented on tapdb/sqlc/migrations/000020_asset_unique_key.up.sql:
```diff
@@ -1,4 +1,107 @@
+DROP INDEX IF EXISTS asset_witnesses_asset_id_witness_index_unique;
+
+-- Step 5: Update the asset_witnesses and asset_proofs tables to reference the
+-- new asset_ids. But skip any records that already reference the new asset_id.
```
I think I would be more confident in this step if we didn't do "skip any records that already reference the new asset_id", which I think is unnecessary. Can't we just set again, value shouldn't change.

guggero replied:
Hmm, you're right. Since we have disabled/removed the unique constraint on the witness table, we can remove the `AND NOT EXISTS` for that table, since it doesn't matter if we create any collisions.

Unfortunately the `asset_proofs` table has an anonymous/inline `UNIQUE` constraint that we cannot disable or remove. So we need to keep that `AND NOT EXISTS` to skip re-assigning proofs if the asset we're going to keep already has one. Otherwise we run into a unique key constraint violation error.

guggero replied:
For the proofs, we have the `AND NOT EXISTS`. So it only moves proofs for assets where there is no proof already for `new_asset_id`. If I remove the `AND NOT EXISTS`, I get a `UNIQUE constraint failed: asset_proofs.asset_id` error.

gijswijs commented on tapdb/testdata/migrations_test_00020_dummy_data.sql:
```diff
@@ -0,0 +1,50 @@
+INSERT INTO asset_witnesses VALUES(3,2,X'a1594fc379308b2a209f6d0bdb8602e9f87cf71fc232c69032b9a5fed28f933100000000',X'add7d0d7cc37e58a7c0d8ad40b6904050d2baa25a1829f00689c4b27b524dd04',X'02827d74858d152da1fae12010ad8d3c46b595c2d4480512a6575925424617124f',X'0101cc',NULL, 0);
+INSERT INTO asset_witnesses VALUES(4,2,X'a1594fc379308b2a209f6d0bdb8602e9f87cf71fc232c69032b9a5fed28f933101000000',X'add7d0d7cc37e58a7c0d8ad40b6904050d2baa25a1829f00689c4b27b524dd04',X'02827d74858d152da1fae12010ad8d3c46b595c2d4480512a6575925424617124f',X'0101dd',NULL, 1);
+
+INSERT INTO asset_proofs VALUES(1,1,X'aaaa');
+
+INSERT INTO asset_proofs VALUES(2,2,X'cccc');
```
Maybe insert only a single proof, linked to the asset we will delete. That way we test the replacement of old ID with new ID with regards to update query for `asset_proofs`

```suggestion
INSERT INTO asset_proofs VALUES(2,2,X'aaaa');
```

guggero replied:
Hmm, you're right. And doing that showed that the query for the proofs isn't correct.
So we need a better query...

gijswijs replied:
I've pushed a commit with a better query.

gijswijs reviewed (commented):
Very nice!! This should fix a lot of the cases of duplication. I think we might run into one or two edge cases where manual intervention is still needed, but the majority (if not all) of it is now covered. I have some questions, a nit and one suggestion that I think will improve the test.

guggero commented on tapdb/sqlc/migrations/000020_asset_unique_key.up.sql:
```diff
@@ -1,4 +1,105 @@
+SET asset_id = (SELECT new_asset_id
+                FROM tmp_asset_id_mapping
+                WHERE old_asset_id = asset_witnesses.asset_id)
+WHERE EXISTS (SELECT 1
```
Hmm, that's way simpler, thanks. Seems to be supported by both engines too.

guggero commented on tapdb/sqlc/migrations/000020_asset_unique_key.up.sql:
```diff
@@ -1,4 +1,105 @@
+WHERE asset_id NOT IN (SELECT min_asset_id FROM tmp_min_assets);
+
+-- Step 9: Delete the duplicates from the assets table. This will then also
+-- delete dangling asset_witnesses, since we have a ON DELETE CASCADE clause on
```
No, since we update all witnesses to point to the asset that we're going to keep, removing the duplicate assets won't remove the now duplicate witnesses. I guess I don't even have to mention DELETE CASCADE here...

jharveyb replied:
Would be good to remove that mention.

jharveyb commented on tapdb/sqlc/migrations/000020_asset_unique_key.up.sql:
```diff
@@ -1,4 +1,99 @@
+-- Step 7: Re-enable the unique index on the asset_witnesses table.
+CREATE UNIQUE INDEX asset_witnesses_asset_id_witness_index_unique
+    ON asset_witnesses (
+                        asset_id, witness_index
```
nit: over-indented.

jharveyb reviewed (commented):
Looks good, just some lingering Qs on query specifics and explanatory comments.

gijswijs commented on tapdb/sqlc/migrations/000020_asset_unique_key.up.sql:
```diff
@@ -1,4 +1,99 @@
+UPDATE asset_proofs
+SET asset_id = filtered_mapping.new_asset_id
+FROM (                  
+        SELECT MIN(old_asset_id) AS old_asset_id, new_asset_id
```
`asset_id` in table `asset_proofs` has a `UNIQUE` constraint. So if we map `old_asset_id` a and b both to `new_asset_id` x we violate that constraint. So by grouping on `new_asset_id` and taking `MIN(old_asset_id)` we only map a single `old_asset_id` to the `new_asset_id`. (a to x in this case) The remaining asset_proof with foreign key b will be deleted in step 8.

The `filtered_mapping` table does the exact same thing as `tmp_asset_id_mapping` but taking into account the `UNIQUE` constraint on `asset_id` in table `asset_proofs`. Hence I reused the column names.

> So the end result would be that, for duplicate assets, if the proof was not already using the new_asset_id (which is the smaller asset ID), asset_proofs.asset_id would be updated to use that new_asset_id.

Correct.

> And any proof already referencing that new_asset_id should be unchanged.

It will get updated, but with the same values, so the result is unchanged.

jharveyb reviewed (approved):
LGTM, great fix and should be safe 👍🏽
